WFLY-6634 Add remote-cache-container configuration to Infinispan subsystem

rhusar opened the pull request:
…finispan subsystem

Jira
https://issues.jboss.org/browse/WFLY-6634

jmesnil commented:
You'll need to create a PR against the EAP 7.x branch that references the corresponding JBEAP issue to get it merged.

rhusar commented:
Hm, it's not clear yet if this is for EAP 7.1 or 7.2 so no downstream PR until then, but I will remove the [11.x] from the title, as that might be confusing. Unfortunately that's the highest number we have in Jira right now.

pferraro commented on clustering/infinispan/extension/src/main/resources/schema/jboss-as-infinispan_5_0.xsd:
```diff
@@ -0,0 +1,1380 @@
+                </xs:documentation>
+            </xs:annotation>
+        </xs:attribute>
+        <xs:attribute name="lifo" type="xs:boolean" default="true">
```
You know my general philosophy on boolean attributes...
What do you think of replacing this with strategy="FIFO|LIFO" or pool="QUEUE|STACK"?

tristantarrant commented on clustering/infinispan/extension/src/test/resources/org/jboss/as/clustering/infinispan/subsystem/subsystem-infinispan-5_0.xml:
```diff
@@ -0,0 +1,148 @@
+                            max-retries="2"
+                            name="my-remote-container"
+                            protocol-version="2.5"
+                            server-socket-bindings="jdg1 jdg2 jdg3"
```
Since we support failover clusters, the server configuration should be extracted to its own sub-element.
<remote-cache-container>
    <remote-servers>
       <remote-server socket-bindings="jdg1 jdg2 jdg3"/>
       <remote-server cluster-name="failoversite" socket-bindings="jdg4 jdg5 jdg6"/>
    </remote-servers>
</remote-cache-container>

tristantarrant replied:
My preference is for the schema to support a single <remote-server> initially with the possibility in the future to support multiples.
Probably instead of remote-server (which conveys the concept of a single node) this should be called remote-cluster

pferraro commented on clustering/infinispan/extension/src/test/resources/org/jboss/as/clustering/infinispan/subsystem/subsystem-infinispan-5_0.xml:
```diff
@@ -0,0 +1,148 @@
+                            max-retries="2"
+                            name="my-remote-container"
+                            protocol-version="2.5"
+                            server-socket-bindings="jdg1 jdg2 jdg3"
```
@tristantarrant Our XML parser is versioned - so future proofing is not an issue.

tristantarrant replied:
Sure, ours is too, but I'd rather future proof the design now.

rhusar committed:
WFLY-6634 Add remote-cache-container/remote-cache configuration to Infinispan subsystem
WFLY-6634 Incorporate feedback for remote-cache-container support
WFLY-6634 Rework connection pool configuration into strategy="FIFO|LIFO"
WFLY-6634 Drop "force-return-values" as a per cache container configuration
WFLY-6634 Don't keep ConfigurationBuilder instances
WFLY-6634 Expose switchToCluster() mechanism methods
WFLY-6634 Introduce support for hotrod failover clusters
WFLY-6634 Merge separate thread pool resource into ThreadPoolResourceDefinition#CLIENT; remove unnecessary model checks in the writer; other cleanup
WFLY-9613 Reduce lambda usage (clustering - remote cache)
Get rid of couple of compiler warnings

jfclere commented:
@jmesnil what about the rebase-this flag, it looks good now.

jmesnil commented:
@jfclere I removed the label

pferraro commented on clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/ConnectionPoolBuilder.java:
```diff
@@ -0,0 +1,96 @@
+    public ConnectionPoolConfiguration getValue() throws IllegalStateException, IllegalArgumentException {
+        return new ConfigurationBuilder().connectionPool()
+                .exhaustedAction(this.exhaustedAction)
+                .lifo(ConnectionPoolStrategy.LIFO.equals(this.connectionPoolStrategy))
```
Use == for enums.

rhusar replied:
Fixed.

pferraro commented on clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLWriter.java:
```diff
@@ -141,6 +142,64 @@ public void writeContent(XMLExtendedStreamWriter writer, SubsystemMarshallingCon
+                        writeThreadPoolElements(XMLElement.ASYNC_THREAD_POOL, ClientThreadPoolResourceDefinition.ASYNC, writer, remoteContainer);
+                    }
+
+                    if (remoteContainer.hasDefined(ConnectionPoolResourceDefinition.PATH.getKeyValuePair())) {
```
This is a required child, so this resource is guaranteed to exist.

rhusar replied:
Fixed, removed.

pferraro commented on clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLWriter.java:
```diff
@@ -141,6 +142,64 @@ public void writeContent(XMLExtendedStreamWriter writer, SubsystemMarshallingCon
+
+//                    if (remoteContainer.hasDefined(RemoteClustersResourceDefinition.PATH.getKeyValuePair())) {
+//                        ModelNode remoteClusters = remoteContainer.get(RemoteClustersResourceDefinition.PATH.getKeyValuePair());
+//                        writeAttributes(writer, remoteClusters, EnumSet.allOf(RemoteClustersResourceDefinition.Attribute.class));
```
Please delete these 3 lines.

rhusar replied:
Oops, removed.

pferraro commented on clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLWriter.java:
```diff
@@ -141,6 +142,64 @@ public void writeContent(XMLExtendedStreamWriter writer, SubsystemMarshallingCon
+//                        ModelNode remoteClusters = remoteContainer.get(RemoteClustersResourceDefinition.PATH.getKeyValuePair());
+//                        writeAttributes(writer, remoteClusters, EnumSet.allOf(RemoteClustersResourceDefinition.Attribute.class));
+
+                        if (remoteContainer.hasDefined(RemoteClusterResourceDefinition.WILDCARD_PATH.getKey())) {
```
There will always be at least 1 remote cluster, so this check is unnecessary.

rhusar replied:
Right, the original idea was to support the collocated cluster. Removed now.

pferraro commented on clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLWriter.java:
```diff
@@ -141,6 +142,64 @@ public void writeContent(XMLExtendedStreamWriter writer, SubsystemMarshallingCon
+
+                    writeAttributes(writer, remoteContainer, EnumSet.allOf(RemoteCacheContainerResourceDefinition.Attribute.class));
+
+                    if (remoteContainer.hasDefined(ClientThreadPoolResourceDefinition.WILDCARD_PATH.getKey())) {
```
This is a required child, so this resource is guaranteed to exist.

rhusar replied:
Fixed, removed.

pferraro commented on clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/ThreadPoolDefinition.java:
```diff
@@ -37,4 +40,8 @@
 
     Attribute getKeepAliveTime();
+
+    PathElement getPathElement();
```
Why is this necessary?

rhusar replied:
Fixed, removed.

rhusar replied:
Originally needed to support both thread pool definitions in generic writer method.

pferraro commented on clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/ThreadPoolDefinition.java:
```diff
@@ -37,4 +40,8 @@
+
+    PathElement getPathElement();
+
+    Collection<Attribute> getAttributes();
```
Why is this necessary?

rhusar replied:
Fixed, removed.

rhusar replied:
Originally needed to support both thread pool definitions in generic writer method.

pferraro commented on clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/ClientThreadPoolResourceDefinition.java:
```diff
@@ -0,0 +1,147 @@
+ *
+ * @author Radoslav Husar
+ */
+public enum ClientThreadPoolResourceDefinition implements ResourceDefinitionProvider, ThreadPoolDefinition {
```
Why create another resource definition?  Why not just add a CLIENT value to the existing ThreadPoolResourceDefinition enum?

rhusar replied:
Merged into ThreadPoolResourceDefinition#CLIENT now.

pferraro commented on clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/ClientThreadPoolBuilder.java:
```diff
@@ -0,0 +1,90 @@
+/**
+ * @author Radoslav Husar
+ */
+public class ClientThreadPoolBuilder extends ComponentBuilder<ExecutorFactoryConfiguration> implements ResourceServiceBuilder<ExecutorFactoryConfiguration> {
```
Can we not just reuse the existing ThreadPoolBuilder?

rhusar replied:
This is needed because it builds diffent objects: org.infinispan.client.hotrod.configuration.ExecutorFactoryConfiguration as opposed to org.infinispan.configuration.global.ThreadPoolConfiguration....

pferraro commented on clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/ThreadPoolDefinition.java:
```diff
@@ -37,4 +37,5 @@
     Attribute getQueueLength();
 
     Attribute getKeepAliveTime();
+
```
?

pferraro commented on clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerResourceDefinition.java:
```diff
@@ -0,0 +1,155 @@
+        new InvalidationNearCacheResourceDefinition().register(registration);
+        new NoNearCacheResourceDefinition().register(registration);
+
+        EnumSet.of(ThreadPoolResourceDefinition.CLIENT).forEach(p -> p.register(registration));
```
ThreadPoolResourceDefinition.CLIENT.register(registration);

pferraro commented on clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteClusterResourceDefinition.java:
```diff
@@ -0,0 +1,151 @@
+
+
+    enum Capability implements org.jboss.as.clustering.controller.Capability {
+        REMOTE_CLUSTER("org.wildfly.clustering.infinispan.remote-cache-container.remote-cluster", Void.class),
```
Make sure to add this to wildfly-capabilities documentation.

rhusar replied:
PR open https://github.com/wildfly/wildfly-capabilities/pull/20

pferraro commented on clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteConfigurationBuilder.java:
```diff
@@ -0,0 +1,177 @@
+
+    @Override
+    public Builder<Configuration> configure(OperationContext context, ModelNode model) throws OperationFailedException {
+        this.connectionTimeout = Attribute.CONNECTION_TIMEOUT.getDefinition().resolveModelAttribute(context, model).asInt();
```
You don't need to call getDefinition() - this method is exposed via the Attribute interface.

rhusar replied:
Done, removed in all other places.

pferraro commented on clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerBuilder.java:
```diff
@@ -0,0 +1,96 @@
+
+        Service<RemoteCacheContainer> service = new SuppliedValueService<>(mapper, supplier, destroyer);
+        ServiceBuilder<RemoteCacheContainer> builder = target.addService(this.getServiceName(), service)
+                .setInitialMode(ServiceController.Mode.PASSIVE)
```
This should be ON_DEMAND.

pferraro commented on clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteConfigurationBuilder.java:
```diff
@@ -0,0 +1,177 @@
+        ServiceBuilder<Configuration> builder = target.addService(this.getServiceName(), new ValueService<>(this))
+                .addDependency(RemoteCacheContainerComponent.CONNECTION_POOL.getServiceName(this.address), ConnectionPoolConfiguration.class, this.connectionPool)
+                .addDependency(RemoteCacheContainerComponent.NEAR_CACHE.getServiceName(address), NearCacheConfiguration.class, this.nearCache)
+                .setInitialMode(ServiceController.Mode.PASSIVE)
```
This should be ON_DEMAND.

rhusar replied:
Done. Original when the PR was open we were still installing services as passive.

pferraro commented on clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/ClientThreadPoolBuilder.java:
```diff
@@ -0,0 +1,91 @@
+        int queueLength = this.definition.getQueueLength().resolveModelAttribute(context, model).asInt();
+        long keepAliveTime = this.definition.getKeepAliveTime().resolveModelAttribute(context, model).asLong();
+
+        ExecutorFactory factory = new DefaultAsyncExecutorFactory() {
```
No need to extend DefaultAsyncExecutorFactory.

pferraro commented on clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/ClientThreadPoolBuilder.java:
```diff
@@ -0,0 +1,91 @@
+            public ExecutorService getExecutor(Properties p) {
+                ThreadFactory clThreadFactory = new ClassLoaderThreadFactory(runnable -> {
+                    Thread thread = new Thread(runnable, THREAD_NAME + "-" + counter.getAndIncrement());
+                    thread.setDaemon(true);
```
Just a TODO - figure out why Infinispan wants these to be daemon threads.

tristantarrant commented:
As far as the schema is specified, I'm happy with it.

rhusar commented:
Thanks Tristan!

rhusar commented:
@jmesnil This is ready to go for a while now, and since you are the only one with merge rights who have commented on the PR and also since you have indicated on other PR to hold on Infinispan 9 upgrade until Artemis upgrade, it looks like this PR should go first. Would you mind looking into it? Thanks!

jmesnil commented:
@rhusar I mentioned to hold changes in messaging-activemq subsystem until Artemis 2.4 is released in #10678 but this PR was for upgrading JGroups, not Infinispan.
Afaik, there is no relation between Infinispan and the messaging-activemq subsystem so I have no issue with merging that PR.

rhusar commented:
@jmesnil Just to clarify this: the Infinispan upgrade *depends* on the JGroups upgrade. So there is now a pretty long dependency chain which this PR is not a part of and can be merged in the meantime.

jmesnil commented:
This PR was on hold until Artemis upgrade was done.
But this is taking longer than expected and I see no reason to hold this PR as @pferraro has approved the changes for it.

jmesnil commented:
retest this please

kabir commented:
@rhusar As this pull request has  been held up for a long time, I am closing it to get the pull request queue under control again. Since this is very old now, it will be a very long time until this one will be addressed. Also, I don't have much time to spend on this to evaluate on a case by case basis.

If this is something important to you or the project please feel free to recreate the pull request. That way it will come on top of the pile again, and be looked at again sooner. 

If it is a new **feature**, please follow the new process for features. Ping @kabir or @jamez if you are unsure what this means, and we can help. Please do that before opening the pull request though.

Sorry for any inconvenience!

rhusar commented:
@kabir This is depressing.

kabir commented:
I see this is a feature now :)

rhusar committed:
WFLY-6634 Adapt remote cache container support for Infinispan 9.1

rhusar committed:
WFLY-9613 Reduce lambda usage (clustering - remote cache) and related cleanup

rhusar committed:
Introduce new Infinispan schema 6.0 and model 7.0.0

rhusar committed:
WFLY-6634 Adapt remote cache container support for Infinispan 9.2

pferraro commented on clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerConfigurationBuilder.java:
```diff
@@ -0,0 +1,187 @@
+
+    @Override
+    public Configuration getValue() {
+        ConfigurationBuilder builder = new ConfigurationBuilder()
```
Additionally, we need to configure the Marshaller.  Similiar to what is done here:
https://github.com/wildfly-clustering/wildfly-clustering-web-hotrod/blob/master/src/main/java/org/wildfly/clustering/web/hotrod/HotRodMarshaller.java
Except that it should use a ModularClassResolver.

rhusar replied:
Done; added new commit to make review easier.

pferraro commented on clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerResourceDefinition.java:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    public enum Attribute implements org.jboss.as.clustering.controller.Attribute {
+        CONNECTION_TIMEOUT("connection-timeout", ModelType.INT, new ModelNode(60000)),
```
We need to add a "module" attribute used to configure the marshaller.

rhusar replied:
Done; added new commit to make review easier.

rhusar committed:
WFLY-6634 Add support for marshaller configuration

pferraro commented on clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteCacheContainerResourceDefinition.java:
```diff
@@ -0,0 +1,157 @@
+
+    public enum Attribute implements org.jboss.as.clustering.controller.Attribute {
+        CONNECTION_TIMEOUT("connection-timeout", ModelType.INT, new ModelNode(60000)),
+        DEFAULT_REMOTE_CLUSTER("default-remote-cluster", ModelType.STRING, builder -> builder.setRequired(true).setCapabilityReference(new CapabilityReference(Capability.CONFIGURATION, RemoteClusterResourceDefinition.Requirement.REMOTE_CLUSTER))),
```
Please replace these lambdas using the strategy employed here: https://github.com/wildfly/wildfly/pull/11079

rhusar committed:
WFLY-9940 Eliminate lambdas from clustering subsystem Attribute definitions

rhusar committed:
WFLY-6634 Adapt remote cache container support for Infinispan 9.2.1

rhusar committed:
Use DynamicCapabilityNameResolver#PARENT_CHILD

rhusar commented on clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteClusterResourceDefinition.java:
```diff
@@ -0,0 +1,153 @@
+        }
+
+        @Override
+        public RuntimeCapability<Void> resolve(PathAddress address) {
```
Done.

pferraro commented on clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/remote/RemoteClusterResourceDefinition.java:
```diff
@@ -0,0 +1,148 @@
+        }
+    }
+
+     enum Requirement implements BinaryRequirement, BinaryServiceNameFactoryProvider {
```
looks like an extra space here?

rhusar replied:
Well spotted, fixed thanks.

rhusar committed:
WFLY-10269 remote-cache-container should use unresolved hostnames to accomodate dynamic environments

rhusar committed:
Adapt to BinaryCapabilityNameResolver.PARENT_CHILD
